lib/logutils/logutils.go

```diff
@@ -0,0 +1,430 @@
+// Copyright (c) 2016-2017 Tigera, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package logutils
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"log/syslog"
+	"os"
+	"path"
+	"runtime"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/mipearson/rfw"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// FilterLevels returns all the logrus.Level values <= maxLevel.
+func FilterLevels(maxLevel log.Level) []log.Level {
+	levels := []log.Level{}
+	for _, l := range log.AllLevels {
+		if l <= maxLevel {
+			levels = append(levels, l)
+		}
+	}
+	return levels
+}
+
+// Formatter is our custom log formatter, which mimics the style used by the Python version of
```

Comment:
This comment is probably less helpful than when I wrote it since there's no Python version of Felix any more.  How about adding a bit of design exposition so that people following know why it's done the way it's done:
```
Formatter is a custom log formatter designed to balance ease of machine processing with human readability.  It includes:

    - A sortable millisecond timestamp, which makes it easier to scan and correlate logs.
    - The level, up front, to aid visual scanning.
    - The PID of the process, to make it easier to spot log discontinuities (If you're looking at two disjoint chunks of log, were they written by the same process?  Was there a restart in-between?)
    - The file name and line number, as essential context.
    - The message!
    - Log fields appended in sorted order.

... example ...

Note: there are tools in Calico's CI pipeline that reply on being able to parse (at lest the prefix section of) this log format in order to detect errors.  If you're tempted to change it, make sure you track down those tools and adjust them accordingly!
```